main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala

```diff
@@ -0,0 +1,208 @@
+      case class RecordLabelPattern(label: Name.Label, pat: Pattern, tpe: Type, loc: SourceLocation)
+    }
+  }
+
+  case class StructField(sym: Symbol.StructFieldSym, tpe: Type, loc: SourceLocation)
+
+  case class JvmMethod(ident: Name.Ident, fparams: List[MonoAst.FormalParam], exp: Expr, retTpe: Type, eff: Type, loc: SourceLocation)
+
+  case class CatchRule(sym: Symbol.VarSym, clazz: java.lang.Class[?], exp: Expr)
+
+  case class HandlerRule(op: OpSymUse, fparams: List[MonoAst.FormalParam], exp: Expr)
+
+  case class MatchRule(pat: Pattern, guard: Option[Expr], exp: Expr)
+
+  case class TypeMatchRule(sym: Symbol.VarSym, tpe: Type, exp: Expr)
+
+  case class SelectChannelRule(sym: Symbol.VarSym, chan: Expr, exp: Expr)
+
+  case class ParYieldFragment(pat: Pattern, exp: Expr, loc: SourceLocation)
+
+  /**
+    * Represents occurrence information of binders, i.e., how a binder occurs in the program.
+    * A binder may be a variable, function, or effect handler.
+    */
+  sealed trait Occur
+
+  object Occur {
+
+    /**
+      * Represents a binder that is not used in an expression.
+      */
+    case object Dead extends Occur
+
+    /**
+      * Represents a binder that occurs exactly once in an expression.
+      */
+    case object Once extends Occur
+
+    /**
+      * Represents a binder that occurs exactly once and that occurrence is in the body of a lambda abstraction.
+      */
+    case object OnceInLambda extends Occur
+
+    /**
+      * Represents a binder that occurs exactly once and that occurrence is in the body of a local def.
+      */
+    case object OnceInLocalDef extends Occur
+
+    /**
+      * Represents a binder that occurs in more than one branch, e.g., match cases, but never inside a lambda abstraction or local def.
+      */
+    case object ManyBranch extends Occur
+
+    /**
+      * Represents a binder that occurs more than once (including lambdas, local defs, branches).
+      */
+    case object Many extends Occur
+
+    /**
+      * Represents a binder that is excluded from inlining at its occurrence sites.
+      * If the binder is a function, sub-expressions of the body may be considered for rewriting.
+      */
+    case object DontInline extends Occur
+
+    /**
+      * Represents a binder that is excluded from inlining at its occurrence sites.
+      * Unlike [[DontInline]], the body of a function is never considered for rewriting (due to casts being dangerous).
+      */
+    case object Dangerous extends Occur
+  }
+
+  /**
+    * A [[DefContext]] contains various pieces information on a function that are relevant for making an inlining decision.
+    *
+    * @param occur           the occurrence information of the function (as described by [[Occur]]).
+    * @param size            the number of subexpressions in the function.
+    * @param localDefs       the number of local defs defined in a function.
+    * @param isDirectCall    true if the outermost expression of the body is a function call.
```

Comment:
Will your inliner actually use this information? Why where? I thought this was a "cheap trick"?

Reply:
I don't understand. This seems quite useful and is currently used by the inliner.